component/loki/source/file/file_test.go

```diff
@@ -76,6 +76,63 @@ func Test(t *testing.T) {
 	}
 }
 
+func TestFileWatch(t *testing.T) {
+	defer goleak.VerifyNone(t, goleak.IgnoreTopFunction("go.opencensus.io/stats/view.(*worker).start"))
+	ctx, cancel := context.WithCancel(componenttest.TestContext(t))
+
+	// Create file to log to.
+	f, err := os.CreateTemp(t.TempDir(), "example")
+	require.NoError(t, err)
+	defer f.Close()
+
+	ctrl, err := componenttest.NewControllerFromID(util.TestLogger(t), "loki.source.file")
+	require.NoError(t, err)
+
+	ch1 := loki.NewLogsReceiver()
+
+	args := Arguments{
+		Targets: []discovery.Target{{
+			"__path__": f.Name(),
+			"foo":      "bar",
+		}},
+		ForwardTo: []loki.LogsReceiver{ch1},
+		FileWatch: FileWatch{
+			MinPollFrequency: time.Millisecond * 500,
+			MaxPollFrequency: time.Millisecond * 500,
+		},
+	}
+
+	go func() {
+		err := ctrl.Run(ctx, args)
+		require.NoError(t, err)
+	}()
+
+	ctrl.WaitRunning(time.Minute)
+
+	timeBeforeWriting := time.Now()
+
+	// Sleep for 600ms to miss the first poll, the next poll should be 2*MinPollFrequency later.
+	time.Sleep(time.Millisecond * 600)
+
+	_, err = f.Write([]byte("writing some text\n"))
+	require.NoError(t, err)
+
+	select {
+	case logEntry := <-ch1.Chan():
+		require.Greater(t, time.Since(timeBeforeWriting), 1*time.Second)
+		require.WithinDuration(t, time.Now(), timeBeforeWriting, 2*time.Second)
+		require.Equal(t, "writing some text", logEntry.Line)
+	case <-time.After(5 * time.Second):
+		require.FailNow(t, "failed waiting for log line")
+	}
+
+	// Shut down the component.
+	cancel()
+
+	// Wait to make sure that all go routines stopped.
+	time.Sleep(time.Millisecond * 100)
```

Comment:
It seems like you're hitting this line: https://github.com/grafana/tail/blob/master/watch/polling.go#L107 

From what I can tell, the goroutine will exit after at most `MaxPollFrequency`, so can you use that timing instead? A random `100ms` is a bit of a smell and is often quite brittle and can become flaky, because there's no guarantee that it won't take `101ms`. 

We can also make the test a bit faster, e.g. divide all the timings by 2. Helps when trying to test for flakiness using `go test -count 10 github.com/grafana/agent/component/loki/source/file -run TestFileWatch`

Reply:
the problem is that if I would divide all the timings by 2, then I would set 250ms which is the actual default. The test needs to check that the new values are set correctly so that would not test correctly